File: llm_written/claude-sonnet-4-5/gemm/java/GemmParallel.java
import java.util.concurrent.*;
import java.util.ArrayList;
import java.util.List;

/**
 * Parallel implementation of GEMM (General Matrix Multiply).
 * Computes C := alpha * A * B + beta * C using tiled matrix multiplication
 * with deterministic parallelization over output tiles.
 */
public class GemmParallel {

    private static final int SEQUENTIAL_THRESHOLD = 256;
    private static final ForkJoinPool POOL = new ForkJoinPool(
        Runtime.getRuntime().availableProcessors()
    );

    public static int[] getSize(double[][] matrix) {
        return new int[]{matrix.length, matrix[0].length};
    }

    public static void validateMatrix(double[][] matrix, String name) {
        if (matrix == null || matrix.length == 0) {
            throw new IllegalArgumentException(name + " must be a non-empty array of rows.");
        }
        if (matrix[0] == null || matrix[0].length == 0) {
            throw new IllegalArgumentException(name + " must have at least one column.");
        }
        int cols = matrix[0].length;
        for (int r = 0; r < matrix.length; r++) {
            if (matrix[r] == null) {
                throw new IllegalArgumentException(name + " row " + r + " is null.");
            }
            if (matrix[r].length != cols) {
                throw new IllegalArgumentException(name + " is ragged: row " + r + " has " +
                    matrix[r].length + " cols, expected " + cols + ".");
            }
        }
    }

    public static double[][] zeros(int rows, int cols) {
        return new double[rows][cols];
    }

    public static double[][] transpose(double[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        double[][] result = new double[cols][rows];
        for (int j = 0; j < cols; j++) {
            for (int i = 0; i < rows; i++) {
                result[j][i] = matrix[i][j];
            }
        }
        return result;
    }

    public static double[][] packMatrix(double[][] matrix, int a0, int a1, int b0, int b1) {
        int rows = b1 - b0;
        int len = a1 - a0;
        double[][] result = new double[rows][len];
        for (int i = 0; i < rows; i++) {
            System.arraycopy(matrix[b0 + i], a0, result[i], 0, len);
        }
        return result;
    }

    private static void partialMatmul(
            double[][] A,
            double[][] B,
            double[][] C,
            double alpha,
            int m0,
            int n0,
            int kb) {
        for (int iOff = 0; iOff < A.length; iOff++) {
            double[] Aik = A[iOff];
            double[] Ci = C[m0 + iOff];

            for (int jOff = 0; jOff < B.length; jOff++) {
                double[] Bjk = B[jOff];
                double s = 0.0;

                for (int k = 0; k < kb; k++) {
                    s += Aik[k] * Bjk[k];
                }

                Ci[n0 + jOff] += alpha * s;
            }
        }
    }

    /**
     * Task to compute one output tile by accumulating contributions from all k-tiles.
     * Each task is independent and writes to a non-overlapping region of C.
     */
    private static class TileTask implements Callable<Void> {
        private final double[][] A;
        private final double[][] Bt;
        private final double[][] C;
        private final double alpha;
        private final int m0, m1, n0, n1, k, KB;

        public TileTask(double[][] A, double[][] Bt, double[][] C, double alpha,
                        int m0, int m1, int n0, int n1, int k, int KB) {
            this.A = A;
            this.Bt = Bt;
            this.C = C;
            this.alpha = alpha;
            this.m0 = m0;
            this.m1 = m1;
            this.n0 = n0;
            this.n1 = n1;
            this.k = k;
            this.KB = KB;
        }

        @Override
        public Void call() {
            // Sequential accumulation over k-tiles for this output tile
            // This ensures deterministic floating-point accumulation order
            for (int k0 = 0; k0 < k; k0 += KB) {
                int k1 = Math.min(k0 + KB, k);
                double[][] Bpack = packMatrix(Bt, k0, k1, n0, n1);
                double[][] Apack = packMatrix(A, k0, k1, m0, m1);
                int kb = k1 - k0;
                partialMatmul(Apack, Bpack, C, alpha, m0, n0, kb);
            }
            return null;
        }
    }

    /**
     * Sequential fallback for small matrices to avoid task overhead.
     */
    private static double[][] runSequential(
            double[][] A,
            double[][] B,
            double alpha,
            double[][] C,
            double beta,
            int MB,
            int NB,
            int KB) {
        int m = A.length, k = A[0].length, n = B[0].length;
        double[][] Bt = transpose(B);

        for (int n0 = 0; n0 < n; n0 += NB) {
            int n1 = Math.min(n0 + NB, n);

            for (int k0 = 0; k0 < k; k0 += KB) {
                int k1 = Math.min(k0 + KB, k);
                double[][] Bpack = packMatrix(Bt, k0, k1, n0, n1);

                for (int m0 = 0; m0 < m; m0 += MB) {
                    int m1 = Math.min(m0 + MB, m);
                    double[][] Apack = packMatrix(A, k0, k1, m0, m1);

                    int kb = k1 - k0;
                    partialMatmul(Apack, Bpack, C, alpha, m0, n0, kb);
                }
            }
        }

        return C;
    }

    /**
     * Compute C := alpha * A * B + beta * C
     *
     * @param A     m x k matrix
     * @param B     k x n matrix
     * @param alpha Multiplier for A*B
     * @param C     Optional m x n accumulation buffer (created if null)
     * @param beta  Multiplier for existing C (applied only if C is provided)
     * @param MB    Tile size for M dimension
     * @param NB    Tile size for N dimension
     * @param KB    Tile size for K dimension
     * @return m x n result matrix (C)
     */
    public static double[][] run(
            double[][] A,
            double[][] B,
            double alpha,
            double[][] C,
            double beta,
            int MB,
            int NB,
            int KB) {
        validateMatrix(A, "A");
        validateMatrix(B, "B");

        int[] sizeA = getSize(A);
        int m = sizeA[0], k = sizeA[1];
        int[] sizeB = getSize(B);
        int k2 = sizeB[0], n = sizeB[1];

        if (k != k2) {
            throw new IllegalArgumentException("Shape mismatch: A is (" + m + "," + k + "), B is (" + k2 + "," + n + ").");
        }

        if (C == null) {
            C = zeros(m, n);
        } else {
            validateMatrix(C, "C");
            int[] sizeC = getSize(C);
            int m2 = sizeC[0], n2 = sizeC[1];
            if (m2 != m || n2 != n) {
                throw new IllegalArgumentException("C has shape (" + m2 + "," + n2 + "); expected (" + m + "," + n + ").");
            }
            if (beta != 1.0) {
                for (int i = 0; i < m; i++) {
                    for (int j = 0; j < n; j++) {
                        C[i][j] *= beta;
                    }
                }
            }
        }

        if (alpha == 0.0) {
            return C;
        }

        // Use sequential version for small matrices to avoid task overhead
        if (Math.max(m, n) < SEQUENTIAL_THRESHOLD) {
            return runSequential(A, B, alpha, C, beta, MB, NB, KB);
        }

        // Parallel version: create tasks for each output tile
        double[][] Bt = transpose(B);
        List<Callable<Void>> tasks = new ArrayList<>();

        // Create tasks in row-major order for determinism
        for (int m0 = 0; m0 < m; m0 += MB) {
            int m1 = Math.min(m0 + MB, m);
            for (int n0 = 0; n0 < n; n0 += NB) {
                int n1 = Math.min(n0 + NB, n);
                tasks.add(new TileTask(A, Bt, C, alpha, m0, m1, n0, n1, k, KB));
            }
        }

        // Execute all tasks using bounded ForkJoinPool
        try {
            List<Future<Void>> futures = POOL.invokeAll(tasks);
            // Check for exceptions
            for (Future<Void> future : futures) {
                future.get();
            }
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException("Parallel execution failed", e);
        }

        return C;
    }

    // Convenience overloads
    public static double[][] run(double[][] A, double[][] B) {
        return run(A, B, 1.0, null, 0.0, 64, 64, 64);
    }

    public static double[][] run(double[][] A, double[][] B, double alpha) {
        return run(A, B, alpha, null, 0.0, 64, 64, 64);
    }

    public static double[][] run(double[][] A, double[][] B, double alpha, double[][] C, double beta) {
        return run(A, B, alpha, C, beta, 64, 64, 64);
    }
}

File: llm_written/claude-sonnet-4-5/gemm/python/gemm_parallel.py
from typing import List, Tuple, Optional
import numbers
from concurrent.futures import ProcessPoolExecutor
import os


def get_size(matrix: List[List[float]]) -> Tuple[int, int]:
    """Return (rows, cols) of a rectangular matrix."""
    return len(matrix), len(matrix[0])


def validate_matrix(matrix: List[List[float]], name: str = "Matrix"):
    """Validate matrix is a non-empty rectangular list-of-lists of numbers."""
    if matrix is None or not isinstance(matrix, list) or not matrix:
        raise ValueError(f"{name} must be a non-empty list of rows.")
    if not isinstance(matrix[0], list) or not matrix[0]:
        raise ValueError(f"{name} must have at least one column.")
    
    cols = len(matrix[0])
    for r, row in enumerate(matrix):
        if not isinstance(row, list):
            raise ValueError(f"{name} row {r} is not a list.")
        if len(row) != cols:
            raise ValueError(f"{name} is ragged: row {r} has {len(row)} cols, expected {cols}.")
        for c, v in enumerate(row):
            if not isinstance(v, numbers.Real):
                raise ValueError(f"{name}[{r}][{c}] is not a real number: {v!r}")
        

def zeros(rows: int, cols: int) -> List[List[float]]:
    return [[0.0] * cols for _ in range(rows)]


def partial_matmul(
        A: List[List[float]], 
        B: List[List[float]], 
        C: List[List[float]],
        alpha: float,
        m0: int,
        n0: int,
        kb: int,
) -> None:
    """Compute a tile of C using packed A and B tiles."""
    C_loc = C  # localize lookup
    for i_off, Aik in enumerate(A):
        Ci = C_loc[m0 + i_off]
        A_loc = Aik
        for j_off, Bjk in enumerate(B):
            B_loc = Bjk
            s = 0.0
            for k in range(kb):
                s += A_loc[k] * B_loc[k]
            Ci[n0 + j_off] += alpha * s


def transpose(matrix: List[List[float]]) -> List[List[float]]:
    return [list(col) for col in zip(*matrix)]


def pack_matrix(matrix: List[List[float]], a0: int, a1: int, b0: int, b1: int) -> List[List[float]]:
    return [row[a0:a1] for row in matrix[b0:b1]]


def _compute_tile_task(args):
    """
    Worker function to compute one (m0, n0) tile for a given k0 slice.
    Returns a list of updates: [(row_idx, col_idx, value), ...]
    
    This function replicates the exact accumulation logic of the sequential version
    to ensure bit-exact reproducibility.
    """
    A, Bt, alpha, m0, m1, n0, n1, k0, k1, kb = args
    
    # Pack the required slices
    Apack = pack_matrix(A, k0, k1, m0, m1)
    Bpack = pack_matrix(Bt, k0, k1, n0, n1)
    
    # Compute partial results using the EXACT same logic as sequential
    updates = []
    for i_off, Aik in enumerate(Apack):
        A_loc = Aik
        for j_off, Bjk in enumerate(Bpack):
            B_loc = Bjk
            s = 0.0
            # Use the exact same accumulation order as sequential
            for k in range(kb):
                s += A_loc[k] * B_loc[k]
            # Only record non-zero updates to reduce communication overhead
            if s != 0.0:
                updates.append((m0 + i_off, n0 + j_off, alpha * s))
    
    return updates


def gemm(
        A: List[List[float]], 
        B: List[List[float]],
        alpha: float = 1.0, 
        C: Optional[List[List[float]]] = None, 
        beta: float = 0,
        MB: int = 64,
        NB: int = 64,
        KB: int = 64,
):
    """
    Compute C := alpha * A * B + beta * C (parallel version).
    
    This implementation parallelizes the computation across tiles while preserving
    the exact accumulation order of the sequential version for bit-exact reproducibility.
    
    :param A: m x k matrix (list-of-lists)
    :param B: k x n matrix (list-of-lists)
    :param alpha: multiplier for A*B
    :param C: optional m x n accumulation buffer (created if None)
    :param beta: multiplier for existing C (applied only if C is provided)
    :param MB,NB,KB: tile sizes
    :return: m x n result (C) as list-of-lists
    """
    validate_matrix(A, "A")
    validate_matrix(B, "B")

    m, k = get_size(A)
    k2, n = get_size(B)

    if k != k2:
        raise ValueError(f"Shape mismatch: A is ({m},{k}), B is ({k2},{n}).")
    
    if C is None:
        C = zeros(m, n)
    else:
        validate_matrix(C, "C")
        m2, n2 = get_size(C) 
        if m2 != m or n2 != n:
            raise ValueError(f"C has shape ({m2},{n2}); expected ({m},{n}).")
        if beta != 1.0:
            for i in range(m):
                for j in range(n):
                    C[i][j] *= beta  

    if alpha == 0.0:
        return C
    
    # Transpose B once (same as sequential)
    Bt = transpose(B)
    
    # Determine number of workers (cap at CPU count)
    num_workers = min(os.cpu_count() or 1, 8)
    
    # Compute tile counts
    n_tiles = (n + NB - 1) // NB
    m_tiles = (m + MB - 1) // MB
    
    # Threshold for parallel execution: use parallel only if we have enough work
    # For small problems, the overhead of process creation dominates
    min_tiles_for_parallel = num_workers * 2
    total_tiles_per_k = n_tiles * m_tiles
    
    if total_tiles_per_k < min_tiles_for_parallel:
        # Fall back to sequential for small problems
        for n0 in range(0, n, NB):
            n1 = min(n0 + NB, n)
            for k0 in range(0, k, KB):
                k1 = min(k0 + KB, k)
                Bpack = pack_matrix(Bt, k0, k1, n0, n1)
                kb = k1 - k0
                
                for m0 in range(0, m, MB):
                    m1 = min(m0 + MB, m)
                    Apack = pack_matrix(A, k0, k1, m0, m1)
                    partial_matmul(Apack, Bpack, C, alpha, m0, n0, kb)
        return C
    
    # Parallel execution: iterate over k0 sequentially (for deterministic reduction),
    # but parallelize over (n0, m0) pairs
    for k0 in range(0, k, KB):
        k1 = min(k0 + KB, k)
        kb = k1 - k0
        
        # Build task list for this k0 slice in deterministic order
        tasks = []
        for n0 in range(0, n, NB):
            n1 = min(n0 + NB, n)
            for m0 in range(0, m, MB):
                m1 = min(m0 + MB, m)
                tasks.append((A, Bt, alpha, m0, m1, n0, n1, k0, k1, kb))
        
        # Execute tasks in parallel and collect results in order
        with ProcessPoolExecutor(max_workers=num_workers) as executor:
            results = list(executor.map(_compute_tile_task, tasks))
        
        # Apply updates to C in the order tasks were submitted (deterministic)
        for updates in results:
            for row_idx, col_idx, value in updates:
                C[row_idx][col_idx] += value
    
    return C
